File: python/Adapter/turkey.py
from abc import ABC, abstractmethod
from duck import Duck


class Turkey(ABC):
    @abstractmethod
    def gobble(self):
        pass

    @abstractmethod
    def fly(self):
        pass


class WildTurkey(Turkey):
    def __init__(self):
        pass

    def gobble(self):
        print("Gobble gobble")

    def fly(self):
        print("I'm flying a short distsnce")


class TurkeyAdapter(Duck):
    def __init__(self, turkey: Turkey):
        self.turkey: Turkey = turkey

    def quack(self):
        self.turkey.gobble()

    def fly(self):
        # the distance a turkey fly 5 times reaches the distance a duck fly
        for i in range(5):
            self.turkey.fly()

File: c/Adapter/turkey.c
#include "turkey.h"
void WildTurkeyNew(WildTurkey *self)
{
    self->gobble = WildTurkeyGobble;
    self->fly = WildTurkeyFly;
}
void WildTurkeyGobble(WildTurkey *self)
{
    printf("Gobble gobble\n");
}
void WildTurkeyFly(WildTurkey *self)
{
    printf("I'm flying a short distance\n");
}

void TurkeyAdapterNew(TurkeyAdapter *self, Turkey *turkey)
{
    self->turkey = turkey;
    self->quack = TurkeyAdapterQuack;
    self->fly = TurkeyAdapterFly;
}
void TurkeyAdapterQuack(TurkeyAdapter *self)
{
    self->turkey->gobble(self->turkey);
}
void TurkeyAdapterFly(TurkeyAdapter *self)
{
    for (int i = 0; i < 5; i++)
    {
        self->turkey->fly(self->turkey);
    }
}